### PR TITLE
Added an AutofacAggregateFactory to EventFlow.Autofac

### PR DESCRIPTION
**Pull request by JC008**
Added:
-  RegisterType() method to IServiceRegistration to simplify registering by type
- support for parameters in AutofacResolver.Resolve methods using overloads
- UseAutofacAggregateFactory() to EventFlowOptionsAutofacExtensions

### CODE REVIEW

**Comment by JC008** [comment]
Just hang on...

There are three issues related to using the `AutofacAggregateFactory`:

1. Validating `AggregateRoot` components in `AutofacServiceRegistration.ValidateRegistrations`
   Cause: parameters are detected by the Autofac default constructor detector that are not available
   Solution: either a `soft` exception is needed or aggregates should not be checked

2. A cleaner way is needed to register the AggregateRoot components
   Issue: `.RegisterServices(f => f.RegisterType(typeof(....)))` is cumbersome 
   Solution: add an `.AddAggregateRoots()` extension

3. Is an AutofacAggregateFactory really needed?
   Question: why did @rasmus use Activator

File: Source/EventFlow.Autofac.Tests/EventFlow.Autofac.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{EDCD8854-6224-4329-87C2-9ADD7D153070}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>EventFlow.Autofac.Tests</RootNamespace>
+    <AssemblyName>EventFlow.Autofac.Tests</AssemblyName>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
```
Comment (rasmus) [inline Source/EventFlow.Autofac.Tests/EventFlow.Autofac.Tests.csproj]: @JC008 Please set the framework to 4.5.1. Ideally multiple frameworks would be included in the NuGet, but I haven't create a setup for that... yet.

File: Source/EventFlow.Autofac.Tests/EventFlow.Autofac.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
```
Comment (rasmus) [inline Source/EventFlow.Autofac.Tests/EventFlow.Autofac.Tests.csproj]: @JC008 Treat warnings as errors should be set for both release and debug builds

File: Source/EventFlow/Extensions/EventFlowOptionsAggregatesExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                .Where(t => !t.IsAbstract && t.IsClosedTypeOf(typeof(IAggregateRoot<>)));
+            eventFlowOptions.AddAggregateRoots(aggregateRootTypes);
+            return eventFlowOptions;
+        }
+
+        public static EventFlowOptions AddAggregateRoots(
+            this EventFlowOptions eventFlowOptions,
+            params Type[] aggregateRootTypes)
+        {
+            foreach (var t in aggregateRootTypes)
```
Comment (rasmus) [inline Source/EventFlow/Extensions/EventFlowOptionsAggregatesExtensions.cs]: @JC008 Add validation that the types really are aggregate roots

File: Source/EventFlow/Extensions/EventFlowOptionsAggregatesExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                eventFlowOptions.RegisterServices(sr => sr.RegisterType(t));
+            }
+            return eventFlowOptions;
+        }
+
+        public static EventFlowOptions AddAggregateRoots(
+            this EventFlowOptions eventFlowOptions,
+            IEnumerable<Type> aggregateRootTypes)
+        {
+            foreach (var t in aggregateRootTypes)
```
Comment (rasmus) [inline Source/EventFlow/Extensions/EventFlowOptionsAggregatesExtensions.cs]: @JC008 Same here, validate the types.

File: Source/EventFlow.Autofac.Tests/UnitTests/Aggregates/AggregateFactoryTests.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                // Assert
+                a.Id.Should().Be(id);
+                ar.Resolver.Should().BeAssignableTo<IResolver>();
+                ap.Pinger.Should().BeOfType<Pinger>();
+            }
+        }
+
+        public class Pinger
+        {
+            public void Ping()
```
Comment (rasmus) [inline Source/EventFlow.Autofac.Tests/UnitTests/Aggregates/AggregateFactoryTests.cs]: @JC008 This method doesn't seem to be used anywhere

File: Source/EventFlow.Autofac.Tests/UnitTests/Aggregates/AggregateFactoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                .AddAggregateRoots(typeof(AggregateFactoryTests).Assembly)
+                .RegisterServices(f => f.RegisterType(typeof(Pinger)))
+                .CreateResolver())
+            {
+                // Arrange
+                var id = TestId.New;
+                var sut = resolver.Resolve<IAggregateFactory>();
+
+                // Act
+                var a = await sut.CreateNewAggregateAsync<TestAggregate, TestId>(id);
```
Comment (rasmus) [inline Source/EventFlow.Autofac.Tests/UnitTests/Aggregates/AggregateFactoryTests.cs]: @JC008 Split these tests into separate tests

File: Source/EventFlow.Autofac.Tests/UnitTests/Aggregates/AggregateFactoryTests.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            var containerBuilder = new ContainerBuilder();
+            using (var resolver = EventFlowOptions.New
+                .UseAutofacContainerBuilder(containerBuilder)
+                .UseAutofacAggregateFactory()
+                .AddAggregateRoots(typeof(AggregateFactoryTests).Assembly)
+                .RegisterServices(f => f.RegisterType(typeof(Pinger)))
+                .CreateResolver())
+            {
+                // Arrange
+                var id = TestId.New;
```
Comment (rasmus) [inline Source/EventFlow.Autofac.Tests/UnitTests/Aggregates/AggregateFactoryTests.cs]: @JC008 Rename variables to something that reflects the type, e.g. `testAggregateWithResolver` for a type `TestAggregateWithResolver`. Its long, be improves readability slightly.

**Comment by rasmus** [comment]
@JC008 The reason for using a `Activator.CreateInstance` in the aggregate factory as default, is that some argue that the aggregate shouldn't be dependent on services, but take these services in as arguments on each method that needs it. Its subtle, but as its really easy (once you know how) to change this logic.

It might be a good idea to include a aggregate factory that can create aggregates using the resolver and include it in the `EventFlow` NuGet package, but don't use it as default.

**Comment by rasmus** [comment]
@JC008 I think adding the `AddAggregateRoots(...)` sounds like a better idea as well.

**Comment by rasmus** [comment]
Ups, wrong button. Sorry

File: Source/EventFlow.Autofac.Tests/EventFlow.Autofac.Tests.csproj [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      <Project>{11131251-778d-4d2e-bdd1-4844a789bca9}</Project>
+      <Name>EventFlow</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="app.config" />
+    <None Include="packages.config" />
+  </ItemGroup>
+  <Choose>
+    <When Condition="'$(VisualStudioVersion)' == '10.0' And '$(IsCodedUITest)' == 'True'">
```
Comment (rasmus) [inline Source/EventFlow.Autofac.Tests/EventFlow.Autofac.Tests.csproj]: @JC008 Not sure why this section is included.

File: Source/EventFlow/EventFlow.csproj [hunk 9]
```diff
@@ -20,7 +20,7 @@
     <DefineConstants>DEBUG;TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
Comment (rasmus) [inline Source/EventFlow/EventFlow.csproj]: @JC008 Revert this

**Comment by rasmus** [comment]
@JC008 Very good idea and something that should have been obvious to include. I'm pedantic with reviews as I _try_ to keep the EventFlow code base clean.

**Comment by rasmus** [comment]
@JC008 Looking good. Remember to update the release notes and let me know when the PR is ready for final review/merge.

**Pull request merged**